articles/Frontend/Architecture/good-practices.md

```diff
@@ -0,0 +1,1724 @@
+ * @param {String} label
+ * @param {Function} [onClick]
+ * @fires click when the button is clicked
+ */
+function buttonFactory(container, config) {
+    return component({
+            click() {
+                // performs some action
+                activateSomething();
+                
+                /**
+                 * @event click
+                 */
+                this.trigger('click');
+            }
+        })
+        .setTemplate(buttonTpl)
+        .on('init', function onInit() {
+            if (container) {
+                this.render(container);
+            }           
+        })
+        .on('render', function onRender() {
+            this.getElement().on('click', () => this.click());                 
+        })
+        .on('click', () => {
+            if (this.getConfig().onClick) {
+                this.getConfig().onClick.call(this);
+            }   
+        })
+        .init(config);
+}
+
+// we may add a callback to listen to click
+button('.fixture', {label: 'Ok', onClick: () => { ... }});
+
+// or we may listen to an event for the same result
+button('.fixture', {label: 'Ok'})
+    .on('click', () => { ... });
+```
+
+#### Good example: one purpose API
+In the previous example the API was redundant, and did not add any value. The
+following snippet fix that by removing the useless API.
+
+```javascript
+/**
+ * @param {Element|String} container
+ * @param {Object} config
+ * @param {String} label
+ * @fires click when the button is clicked
+ */
+function buttonFactory(container, config) {
+    return component({
+            click() {
+                // performs some action
+                activateSomething();
+                
+                /**
+                 * @event click
+                 */
+                this.trigger('click');
+            }
+        })
+        .setTemplate(buttonTpl)
+        .on('init', function onInit() {
+            if (container) {
+                this.render(container);
+            }           
+        })
+        .on('render', function onRender() {
+            this.getElement().on('click', () => this.click());                 
+        })
+        .init(config);
+}
+
+// Only a single way of reacting to click actions
+button('.fixture', {label: 'Ok'})
+    .on('click', () => { ... });
+```
+
+#### Resources
+- [Coding guide: General rules](coding-guide.md#general-rules)
+
+## Testing
+
+### Prefer design by coding: TDD / BDD
+When writing code, a developer must check the works is going the right way.
+Usually, some manual checks are made aside. But it may become more and more
+time consuming as long as the progress is made. So it is better to automate
+somehow those manual checks. And on convenient way to do so is to write
+unit tests.
+
+A method well describes the process, it is called [Test Driven Development](https://en.wikipedia.org/wiki/Test-driven_development).
+The principle is to write tests aside at the same period of time you are
```

Comment:
not at the same time, it's before (start with the test then make your implementation fulfil it)